carmm/run/aims_calculator.py

```diff
@@ -227,47 +227,32 @@ def get_k_grid(model, sampling_density, verbose=False, simple_reciprocal_space_p
         return None
 
     # These are lattice vectors
-    x_v = model.get_cell()[0]
-    y_v = model.get_cell()[1]
-    z_v = model.get_cell()[2]
+    lattice_v = np.array(model.get_cell())
     # These are lattice parameters
-    x = np.linalg.norm(x_v)
-    y = np.linalg.norm(y_v)
-    z = np.linalg.norm(z_v)
+    lattice_param = np.array([np.linalg.norm(v) for v in lattice_v])
 
     if simple_reciprocal_space_parameters:
         # Simplified reciprocal lattice parameters
-        l_v_x = 2 * math.pi / x
-        l_v_y = 2 * math.pi / y
-        l_v_z = 2 * math.pi / z
+        reciprocal_param = 2 * math.pi / lattice_param
     else:
         # volume of the cell
-        volume = np.dot(x_v, np.cross(y_v, z_v))
+        volume = np.dot(lattice_v[0], np.cross(lattice_v[1], lattice_v[2]))
         # These are reciprocal lattice vectors
-        v_x = np.cross(y_v, z_v) * 2 * math.pi / volume
-        v_y = np.cross(z_v, x_v) * 2 * math.pi / volume
-        v_z = np.cross(x_v, y_v) * 2 * math.pi / volume
+        reciprocal_v = [np.cross(lattice_v[(i + 1) % 3], lattice_v[(i + 2) % 3]) * 2 * math.pi / volume
+                        for i in range(len(lattice_v))]
         # These are reciprocal lattice parameters
-        l_v_x = np.linalg.norm(v_x)
-        l_v_y = np.linalg.norm(v_y)
-        l_v_z = np.linalg.norm(v_z)
+        reciprocal_param = np.array([np.linalg.norm(r_v) for r_v in reciprocal_v])
 
     k_grid_density = 1 / (sampling_density * 2 * math.pi)
-
-    k_x = math.ceil(k_grid_density * l_v_x)
-    k_y = math.ceil(k_grid_density * l_v_y)
-    k_z = math.ceil(k_grid_density * l_v_z)
-
+    k_grid = k_grid_density * reciprocal_param
+    # Convert k_grid to integer
+    k_grid = np.array([math.ceil(k) for k in k_grid])
     # Remove k-sampling if direction is not periodic in any dimension
-    if dimensions < 3:
-        k_z = 1
-    if dimensions < 2:
-        k_y = 1
-
-    k_grid = (k_x, k_y, k_z)
+    k_grid[np.invert(model.pbc)] = 1
```

Comment:
Model.pbc is a Boolean vector of length 3, and invert changes anything that is not periodic to True, and then sets it to 1. This is all good, I’m just puzzled by the fact the assignment is done within the k_grid [] rather than `for k in model.pbc`. This is probably my naivety.

A bigger problem is we rely on `model.pbc` here to set dimensionality, but in the other functions we just pass in `dimension` as a user input. I think there should be some consistency - my gut feeling is we should make it user driven, because some users might not set `model.pbc` as they don’t realise it is needed?

Reply:
Since `k_grid` is now a numpy array, `k_grid[np.invert(model.pbc)] = 1` takes advantage of boolean array indexing of numpy arrays. `for k in model.pbc` will also work though. I think they are both fine.

Yes, dimensions is not an user input in this function (or a previous version of this function). It is calculated using the sum of `pbc` in line 222 `dimensions = sum(model.pbc)`. So I think the implementation here is coherent with existing infrastructure.

Additionally, `pbc` is often set implicitly. For example, when build a slab with ase.build.fcc111, the `pbc` will be set as [ True,  True, False]. If `pbc` is not set at all, the default value would be `False` which would give `dimesions` as 0 and the `get_k_grid` function will stop and print warning. I wanted to use model.pbc here, because it seems more flexible especially in cases where the user want to use "unusual" `pbc` for models such as a wire along y axis or a slab normal to x axis. But I am probably overthinking.

Reply:
I don't think it's overthinking - just making sure we explore all options before finalising! I'm happy with all the above, and IMO we can merge when you have closed any remaining open issues ... If there are any?

Reply:
Thanks Andy! I am waiting for some more comments from Akash.

Reply:
Akash raised some concern with pymatgen-generated slabs this afternoon. After conversion to ase.Atoms, slabs from pymatgen has `pbc [True, True, True]`. If using a big slab, the k grid along reciprocal z axis could be more than 1.
I added a few lines to check if the model is periodic and with vacuum along a certain axis. If so, the code will still proceed, but it will print a message to warn the user. The other option is to set `pbc` to `False` in such a dimension within the function.
I chose the former, because it is possible that this is intended. For example, we use adsorbate in a cell with same periodic condition as the adsorption complex when doing a counterpoise correction for BSSE.